File: firebase/functions/lib/events/live_meetings/breakouts/initiate_breakouts.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'assign_to_breakouts.dart';
import '../../../on_call_function.dart';
import 'check_assign_to_breakouts_server.dart';
import '../../../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/live_meeting.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/utils/utils.dart';
import 'package:meta/meta.dart';

class InitiateBreakouts extends OnCallMethod<InitiateBreakoutsRequest> {
  @visibleForTesting
  static math.Random random = math.Random();

  InitiateBreakouts()
      : super(
          InitiateBreakoutsRequest.functionName,
          (json) => InitiateBreakoutsRequest.fromJson(json),
          runWithOptions: RuntimeOptions(
            timeoutSeconds: 120,
            memory: '4GB',
            minInstances: 0,
          ),
        );

  @override
  Future<void> action(
    InitiateBreakoutsRequest request,
    CallableContext context,
  ) async {
    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    print('checking is authorized');
    await _verifyCallerIsAuthorized(event, context);

    await initiateBreakouts(
      request: request,
      event: event,
      creatorId: context.authUid!,
    );
  }

  Future<void> _verifyCallerIsAuthorized(
    Event event,
    CallableContext context,
  ) async {
    final communityMembershipDoc = await firestore
        .document(
          'memberships/${context.authUid}/community-membership/${event.communityId}',
        )
        .get();

    final membership = Membership.fromJson(
      firestoreUtils
          .fromFirestoreJson(communityMembershipDoc.data.toMap() ?? {}),
    );

    final isAuthorized = event.creatorId == context.authUid || membership.isMod;
    if (!isAuthorized) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }
  }

  Future<void> initiateBreakouts({
    required InitiateBreakoutsRequest request,
    required Event event,
    required String creatorId,
  }) async {
    if (event.isHosted) {
      print('Assigning users to breakouts.');
      await AssignToBreakouts().assignToBreakouts(
        targetParticipantsPerRoom: request.targetParticipantsPerRoom,
        breakoutSessionId: request.breakoutSessionId,
        assignmentMethod:
            request.assignmentMethod ?? BreakoutAssignmentMethod.targetPerRoom,
        includeWaitingRoom: request.includeWaitingRoom,
        event: event,
        creatorId: creatorId,
      );
    } else {
      print('Pinging breakout availability.');
      await _pingBreakoutsAvailability(
        event: event,
        request: request,
      );
    }
  }

  Future<void> _pingBreakoutsAvailability({
    required Event event,
    required InitiateBreakoutsRequest request,
  }) async {
    final breakoutRoomSessionId = request.breakoutSessionId;
    final liveMeetingPath = '${event.fullPath}/live-meetings/${event.id}';

    const smartMatchingWaitTime = Duration(seconds: 30);

    final now = DateTime.now();
    final nowWithoutMilliseconds =
        now.subtract(Duration(milliseconds: now.millisecond));
    final scheduledTime = nowWithoutMilliseconds.add(smartMatchingWaitTime);

    final newlyInitiated = await firestore.runTransaction((transaction) async {
      final liveMeetingDocRef = firestore.document(liveMeetingPath);
      final liveMeetingDoc = await transaction.get(liveMeetingDocRef);
      final liveMeeting = LiveMeeting.fromJson(
        firestoreUtils.fromFirestoreJson(liveMeetingDoc.data.toMap()),
      );
      if (liveMeeting.currentBreakoutSession?.breakoutRoomSessionId ==
          breakoutRoomSessionId) {
        print('Breakout session already initiated. Returning');
        return false;
      }
      final breakoutSession = BreakoutRoomSession(
        breakoutRoomSessionId: breakoutRoomSessionId,
        breakoutRoomStatus: BreakoutRoomStatus.pending,
        assignmentMethod: request.assignmentMethod!,
        targetParticipantsPerRoom: request.targetParticipantsPerRoom,
        hasWaitingRoom: request.includeWaitingRoom,
        scheduledTime: scheduledTime,
      );
      transaction.set(
        liveMeetingDocRef,
        DocumentData.fromMap(
          jsonSubset(
            [LiveMeeting.kFieldCurrentBreakoutSession],
            firestoreUtils.toFirestoreJson(
              LiveMeeting(
                currentBreakoutSession: breakoutSession,
              ).toJson(),
            ),
          ),
        ),
        merge: true,
      );
      return true;
    });

    if (newlyInitiated) {
      print('scheduling assign to breakouts server check');
      await CheckAssignToBreakoutsServer().schedule(
        CheckAssignToBreakoutsRequest(
          eventPath: event.fullPath,
          breakoutSessionId: breakoutRoomSessionId,
        ),
        scheduledTime,
      );
    } else {
      print(
        'not enqueuing assign to breakouts checks since it was already setup.',
      );
    }
  }
}

File: firebase/functions/src/events/live_meetings/breakouts/initiate_breakouts.ts
import * as functions from 'firebase-functions'
import { AssignToBreakouts } from './assign_to_breakouts'
import { OnCallMethod } from '../../../on_call_function'
import { firestore, firestoreUtils } from '../../../utils/infra/firestore_utils'
import { Event, Membership, BreakoutAssignmentMethod, membershipIsMod } from '../../../types'

interface InitiateBreakoutsRequest {
    eventPath: string
    breakoutSessionId: string
    assignmentMethod?: BreakoutAssignmentMethod
    targetParticipantsPerRoom: number
    includeWaitingRoom?: boolean
}

export class InitiateBreakouts extends OnCallMethod<InitiateBreakoutsRequest> {
    constructor() {
        super('InitiateBreakouts', (json) => json as InitiateBreakoutsRequest, {
            runWithOptions: { timeoutSeconds: 120, memory: '4GB', minInstances: 0 },
        })
    }

    async action(
        request: InitiateBreakoutsRequest,
        context: functions.https.CallableContext
    ): Promise<void> {
        const event: Event = await firestoreUtils.getFirestoreObject({
            path: request.eventPath,
            constructor: (map) => map as unknown as Event,
        })

        await this._verifyCallerIsAuthorized(event, context)

        await this.initiateBreakouts({ request, event, creatorId: context.auth!.uid })
    }

    private async _verifyCallerIsAuthorized(
        event: Event,
        context: functions.https.CallableContext
    ): Promise<void> {
        const membershipSnap = await firestore
            .doc(
                `memberships/${context.auth?.uid}/community-membership/${
                    (event as any).communityId
                }`
            )
            .get()
        const membership = firestoreUtils.fromFirestoreJson(
            membershipSnap.data() ?? {}
        ) as unknown as Membership

        const isAuthorized =
            (event as any).creatorId === context.auth?.uid ||
            membershipIsMod((membership as any).status)
        if (!isAuthorized) {
            throw new functions.https.HttpsError('failed-precondition', 'unauthorized')
        }
    }

    async initiateBreakouts({
        request,
        event,
        creatorId,
    }: {
        request: InitiateBreakoutsRequest
        event: Event
        creatorId: string
    }): Promise<void> {
        if ((event as any).isHosted || (event as any).eventType === 'hosted') {
            console.log('Assigning users to breakouts.')
            await new AssignToBreakouts().assignToBreakouts({
                targetParticipantsPerRoom: request.targetParticipantsPerRoom,
                breakoutSessionId: request.breakoutSessionId,
                assignmentMethod:
                    request.assignmentMethod ?? BreakoutAssignmentMethod.targetPerRoom,
                includeWaitingRoom: request.includeWaitingRoom ?? false,
                event,
                creatorId,
            })
        } else {
            console.log('Pinging breakout availability.')
            await this._pingBreakoutsAvailability({ event, request })
        }
    }

    private async _pingBreakoutsAvailability({
        event,
        request,
    }: {
        event: Event
        request: InitiateBreakoutsRequest
    }): Promise<void> {
        const breakoutRoomSessionId = request.breakoutSessionId
        const liveMeetingPath = `${(event as any).fullPath}/live-meetings/${(event as any).id}`

        const smartMatchingWaitMs = 30_000
        const nowMillis = Date.now()
        const scheduledMs = Math.floor((nowMillis + smartMatchingWaitMs) / 1000) * 1000
        const scheduledTime = new Date(scheduledMs)

        const newlyInitiated = await firestore.runTransaction(async (transaction) => {
            const liveMeetingRef = firestore.doc(liveMeetingPath)
            const liveMeetingDoc = await transaction.get(liveMeetingRef)
            const liveMeeting = firestoreUtils.fromFirestoreJson(liveMeetingDoc.data() ?? {})

            if (
                (liveMeeting as any).currentBreakoutSession?.breakoutRoomSessionId ===
                breakoutRoomSessionId
            ) {
                console.log('Breakout session already initiated. Returning')
                return false
            }

            const breakoutSession = {
                breakoutRoomSessionId,
                breakoutRoomStatus: 'pending',
                assignmentMethod: request.assignmentMethod,
                targetParticipantsPerRoom: request.targetParticipantsPerRoom,
                hasWaitingRoom: request.includeWaitingRoom,
                scheduledTime,
            }

            transaction.set(
                liveMeetingRef,
                {
                    currentBreakoutSession: firestoreUtils.toFirestoreJson(
                        breakoutSession as Record<string, unknown>
                    ),
                },
                { merge: true }
            )
            return true
        })

        if (newlyInitiated) {
            console.log('scheduling assign to breakouts server check')
            const { CheckAssignToBreakoutsServer } = await import(
                './check_assign_to_breakouts_server'
            )
            await new CheckAssignToBreakoutsServer().schedule(
                { eventPath: (event as any).fullPath, breakoutSessionId: breakoutRoomSessionId },
                scheduledTime
            )
        }
    }
}
